cmd/bundle/exec.go

```diff
@@ -0,0 +1,152 @@
+			// If that's the case, i.e. no targets are defined, then do not pass the target.
+			//
+			// This is only useful for when the Databricks CLI is the child command.
+			if b.Config.Bundle.Target != mutator.DefaultTargetPlaceholder {
+				env = append(env, "DATABRICKS_BUNDLE_TARGET="+b.Config.Bundle.Target)
+			}
+
+			// If the bundle has a profile configured, explicitly pass it to the child command.
+			//
+			// This is only useful for when the Databricks CLI is the child command,
+			// since if we do not explicitly pass the profile, the CLI will use the
+			// auth configured in the bundle YAML configuration (if any).
+			if b.Config.Workspace.Profile != "" {
+				env = append(env, "DATABRICKS_CONFIG_PROFILE="+b.Config.Workspace.Profile)
+			}
+
+			childCmd.Env = env
+
+			// Execute all scripts from the bundle root directory. This behavior can
+			// be surprising in isolation, but we do it to keep the behavior consistent
+			// for both these cases:
+			// 1. One shot commands like `databricks bundle exec -- echo hello`
+			// 2. (upcoming) Scripts that are defined in the scripts section of the DAB.
+			//
+			// TODO(shreyas): Add a DATABRICKS_BUNDLE_INITIAL_CWD environment variable
+			// that users can read to figure out the original CWD. I'll do that when
+			// adding support for the scripts section.
+			childCmd.Dir = b.BundleRootPath
+
+			stdout, err := childCmd.StdoutPipe()
+			if err != nil {
+				return fmt.Errorf("creating stdout pipe failed: %w", err)
+			}
+
+			stderr, err := childCmd.StderrPipe()
+			if err != nil {
+				return fmt.Errorf("creating stderr pipe failed: %w", err)
+			}
+
+			// Start the child command.
+			err = childCmd.Start()
+			if err != nil {
+				return fmt.Errorf("starting %q failed: %w", strings.Join(args, " "), err)
+			}
+
+			var wg sync.WaitGroup
+			wg.Add(2)
+
+			var stdoutErr error
+			go func() {
+				defer wg.Done()
+
+				scanner := bufio.NewScanner(stdout)
+				for scanner.Scan() {
+					_, err = fmt.Fprintln(cmd.OutOrStdout(), scanner.Text())
+					if err != nil {
+						stdoutErr = err
+						break
+					}
+				}
+			}()
+
+			var stderrErr error
+			go func() {
+				defer wg.Done()
+
+				scanner := bufio.NewScanner(stderr)
+				for scanner.Scan() {
+					_, err = fmt.Fprintln(cmd.ErrOrStderr(), scanner.Text())
+					if err != nil {
+						stderrErr = err
+						break
+					}
+				}
+			}()
+
+			wg.Wait()
+
+			if stdoutErr != nil {
+				return fmt.Errorf("writing stdout failed: %w", stdoutErr)
+			}
+
+			if stderrErr != nil {
+				return fmt.Errorf("writing stderr failed: %w", stderrErr)
+			}
+
+			err = childCmd.Wait()
+			if exitErr, ok := err.(*exec.ExitError); ok {
+				// We don't make the parent CLI process exit with the same exit code
+				// as the child process because the exit codes for the CLI have not
+				// been standardized yet.
+				//
+				// This keeps the door open for us to associate specific exit codes
+				// with specific classes of errors in the future.
```

Comment:
I think propagating the code is natural here.

Reply:
That's a one-way door. Once we start doing that we can't take it back. On the other hand we can always start emitting the exit code directly instead of returning `1`.

Right now, we don't have a standardized convention of non-zero exit codes, and we might want to reserve some range of exit codes for the CLI.

Reply:
Sure, we need to think about this carefully. Given that exec is a wrapper around some existing programs, people might expect the code to match. How else would they get this info, e.g. see pytest ones https://docs.pytest.org/en/stable/reference/exit-codes.html

Reply:
I agree that users might expect it. We don't have to support this in v1 though, and can retroactively add it if we decide it's the right thing to do.

It won't be a breaking change because exit code 0 would still map to 0, and non-zero exit codes would be more diverse instead of just being 1 in the CLI parent process.